pyomo/contrib/cspline_external/cspline_parameters.py

```diff
@@ -180,11 +206,18 @@ def segment(self, x):
             array of integers is returned otherwise return an integer
         """
         s = np.searchsorted(self.knots, x)
-        # If x is past the last knot, use the last segment
-        # this could happen just due to round-off even if
-        # you don't intend to extrapolate
-        s[s >= self.n_segments] = self.n_segments - 1
-        return s
+        # If x is before the first knot use the first segment
+        if len(s) > 1:
+            s[s <= 0] = 1
+            s[s >= len(self.knots)] = len(self.knots) - 1
+        else:
+            if s <= 0:
+                return 0
+            if s >= len(self.knots):
+                # if the not extrapolates right it is 1 more than number of
```

Comment:
```suggestion
                # if the knot extrapolates right it is 1 more than number of
```
Is this supposed to be `knot`?